### PR TITLE
Non-trivial sum type exhaustiveness checking

### PR DESCRIPTION
**Pull request by erszcz**
I tried exercising sum type exhaustiveness checking on some practical examples and realised that these checks only worked on simple sum types whose variants were builtin types - e.g. an alternative of atoms:

```erlang
-type allergen() :: eggs
                  | chocolate
                  | pollen
                  | cats.
```

I hoped I could rely on these checks when defining functions over a type with many variants, each carrying extra data. For example:

```erlang
-type term_() :: {true, info()}
               | {false, info()}
               | {'if', info(), term_(), term_(), term_()}
               | {zero, info()}
               | {succ, info(), term_()}
               | {pred, info(), term_()}
               | {is_zero, info(), term_()}.
```

The first approach was to enable `typechecker:refinable()` and `gradualizer_lib:pick_value()` to handle `{user_type, _, _, _}` types. However, this first implementation was prone to falling into an infinite loop on recursive variants. This lead me to add the `Trace` parameter to `refinable` - in other words, I optimistically assume that a type T is refinable as soon as recursion on T's variants reaches T again. This assumes that the other non-recursive variants of T will ultimately determine its refinability. Thanks to storing a trace of already seen types instead of a binary `seen | not-seen` flag mutually recursive types are also handled gracefully.

Then it turned out that if the sum type is defined in another module, the check doesn't work. Obviously, data types like this might be handled in multiple modules of a project, so this was not acceptable. The fix I applied is too look up user types in `gradualizer_db` first and only if not found there try the available `TEnv`. [I tried the reverse order](https://github.com/erszcz/Gradualizer/tree/exhaustive-user-type-normalize-remote-type) (`TEnv` first, then `gradualizer_db`), but it seems to lead to more convoluted code since `#tenv{}` is not available in a header, therefore not shareable with `gradualizer_lib` - nothing impossible to handle, but the former fix was already good enough.

Ultimately, this allows to rely on exhaustiveness warnings when working with (mutually recursive) non-trivial variant types from external modules, which IMO is a significant ergonomics improvement. For a larger example, please see https://github.com/erszcz/tapl-erlang/commit/300d1ed100ab1c820dc2fd54d290345f3449a231 and try:

```
$ rebar3 compile
$ gradualizer -pa _build/default/lib/arith/ebin -- src/arith_core.erl
src/arith_core.erl: Nonexhaustive patterns on line 34 at column 9
Example values which are not covered:
	{false, 0}
```

Gradualizer correctly detects the missing clauses:

```erlang
-type term_() :: arith_syntax:term_().

-spec eval1(term_()) -> term_().
eval1(T) ->
    case T of
        %{false, _} ->
        %    erlang:throw(constant);
        %{true, _} ->
        %    erlang:throw(constant);
        %{zero, _} ->
        %    erlang:throw(constant);
        %{'if', _, {true, _}, T2, _} ->
        %    T2;
        %{'if', _, {false, _}, _, T3} ->
        %    T3;
        %{'if', Info, T1, T2, T3} ->
        %    T1_ = eval1(T1),
        %    {'if', Info, T1_, T2, T3};
        {succ, Info, T1} ->
            T1_ = eval1(T1),
            {succ, Info, T1_};
        {pred, _, {zero, _}} ->
            {zero, 0};
        {pred, _, {succ, _, T1}} ->
            T1;
        {pred, Info, T1} ->
            T1_ = eval1(T1),
            {pred, Info, T1_};
        {is_zero, _, {zero, _}} ->
            {true, 0};
        {is_zero, _, {succ, _, _}} ->
            {false, 0};
        {is_zero, Info, T1} ->
            T1_ = eval1(T1),
            {is_zero, Info, T1_}
    end.
```

### CODE REVIEW

File: src/typechecker.erl
Comment (zuiderkwast) [inline src/typechecker.erl:3460]: Perhaps a matter of taste, and I know it's not even consistent within this module, but I think `refinable(Ty, TEnv)` is a better order of the parameters.

File: src/typechecker.erl
Comment (zuiderkwast) [inline src/typechecker.erl:3480]: No need to check that TEnv is a `#tenv{}` here, is it?

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:111]: I'd prefer to reverse the order of parameters, i.e. put Types param after the Type we're picking a value from. Also add some short comment about what the params are (or a spec) please.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:87]: To get fewer and easier-to-remember parameters, why not keep the `user_type` tuple here? as in

```suggestion
get_user_type_definition({user_type, Anno, Name, Args}, Types) ->
```

A short comment about the function and the parameters is always nice to have. Types is a map from name to type, right? (Or a spec.)

**Review by zuiderkwast — CHANGES_REQUESTED** [review]
Very nice feature!

I have only some suggestions on style.

I have the feeling we now have multiple ways to lookup user types. Is it worth refactoring the other accesses to get less ways of doing it?

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:91]: I just saw you're using `get_opaque_type` here. I don't think want to do exhaustiveness-checking on opaque remote types, since they are opaque...

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:101]: Don't we need to substitute Args for Params in Ty here?

File: src/gradualizer_lib.erl
Comment (erszcz) [reply src/gradualizer_lib.erl]: Indeed, this was just an easy to way to avoid the `opaque` return value from `gradualizer_db:get_type`, which can be handled easily in `refinable` (`opaque` is not refinable), but [will require an error to be thrown if it ever shows up when returned in `pick_value`](https://github.com/erszcz/Gradualizer/blob/exhaustive-user-type/src/gradualizer_lib.erl#L181-L182). I wanted to avoid introducing a new kind of error as I don't see anything relevant to `opaque` in `gradualizer_fmt`.

File: src/gradualizer_lib.erl
Comment (erszcz) [reply src/gradualizer_lib.erl:112]: Hmmm, I'm not sure yet, I'd have to devise a test to work it out on some examples...

**Comment by erszcz** [comment]
I've pushed some changes - all the style / parameter order / spec points are addressed (unless you have more comments on these). The open issues I still have to think of are the `throw(opaque, ...)` from `pick_value` and the `Args <-> Params` substitution - any more feedback is appreciated :)

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [reply src/gradualizer_lib.erl]: Good point. If we want to pick_value of type `{tag, opaque_type()}`, we'd need a value of the opaque type. But OTOH opaque values shouldn't be shown.

Since pick_value is only used for producing error messages, perhaps we can put an underscore or three dots in there instead? `{tag, _}` or `{tag, ...}`, whatever is best. We could also change the exhaustiveness error messages slightly so that it doesn't explicitly say that it's a value of the type. Just some idea. WDYT?

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:94]: Nice. Since this Types map is used in many places, we *could* consider creating a type for this, like `gradualizer:type_map()` or so... But I won't demand it.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:127]: Types might be called UserTypes, UserTypeMap or something (only in the spec) to avoid confusion Ty/Types.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:182]: Idea: return `{var, Anno, '_'}` here, for now. Or generate a unique OpaqueValueNNN variable name using unique_integer or something.

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [reply src/gradualizer_lib.erl:112]: Yes, tests are good to have.

Just thinking that `gradualizer_db:get_type/3` does substitute type params and that local and remote types probably should be handled in the same way.

Comment (zuiderkwast) [reply src/gradualizer_lib.erl:112]: Btw, this clause in `typechecker` is very similar and can maybe be refactored to use `get_user_type_definition`:

```Erlang
normalize({user_type, P, Name, Args} = Type, TEnv) ->
    case typelib:get_module_from_annotation(P) of
        {ok, Module} ->
            %% Local type in another module, from an expanded remote type
            case gradualizer_db:get_type(Module, Name, Args) of
                {ok, T} ->
                    normalize(typelib:remove_pos(T), TEnv);
                opaque ->
                    Type;
                not_found ->
                    throw({undef, user_type, P, {Module, Name, length(Args)}})
            end;
        none ->
            %% Local user-defined type
            TypeId = {Name, length(Args)},
            case TEnv#tenv.types of
                #{TypeId := {Vars, Type0}} ->
                    VarMap = maps:from_list(lists:zip(Vars, Args)),
                    Type1 = typelib:substitute_type_vars(Type0, VarMap),
                    normalize(Type1, TEnv);
                _NotFound ->
                    throw({undef, user_type, P, {Name, length(Args)}})
            end
    end;
```

File: test/should_fail/exhaustive_remote_user_type.erl
Comment (zuiderkwast) [inline test/should_fail/exhaustive_remote_user_type.erl:32]: Really nice that you managed to test this stuff!

File: src/typechecker.erl
Comment (zuiderkwast) [inline src/typechecker.erl:3504]: So now we have two representations of TEnv... Not ideal but I can live with it if there is a plan to unify them in the future.  I'd prefer either of

1. Change TEnv to a map everywhere.
2. Use the record everywhere and move it to an include file.

File: test/test.erl
Comment (zuiderkwast) [inline test/test.erl:38]: You could use `user_types:my_opaque()` in the new test cases. Then you don't need to import another module here.

File: src/typechecker.erl
Comment (erszcz) [reply src/typechecker.erl:3504]: I'd prefer changing it to a map everywhere, unless you think it's better to keep it as a record. I can do this in this PR or a separate one.

Comment (erszcz) [reply src/typechecker.erl:3504]: BTW, this map for the time being is described by https://github.com/erszcz/Gradualizer/blob/exhaustive-user-type/src/gradualizer_lib.erl#L87-L90 - where do you think is the right place to put this type?

File: test/test.erl
Comment (erszcz) [reply test/test.erl:38]: Good idea, I'll try reworking the tests this way :+1:

File: src/typechecker.erl
Comment (zuiderkwast) [reply src/typechecker.erl:3504]: It's internal so any other module than `gradualizer`. I suppose `gradualizer_lib` is fine.

The modules are a bit messy. Some are prefixed by `gradualizer_` and some aren't. The `typechecker` module has `export_all`. We should clean up this mess some day...

Comment (zuiderkwast) [reply src/typechecker.erl:3504]: > I'd prefer changing it to a map everywhere, unless you think it's better to keep it as a record. I can do this in this PR or a separate one.

I guess a separate PR would be better, if it's not too much to ask. It can be done either before or after this PR is merged, as long as it's not forgotten.

File: src/gradualizer_lib.erl
Comment (erszcz) [reply src/gradualizer_lib.erl]: Ok, this one is refactored to use `get_type_definition`, too.

File: src/typechecker.erl
Comment (erszcz) [reply src/typechecker.erl]: Done with #333.

File: test/test.erl
Comment (erszcz) [reply test/test.erl:38]: Thinking about it again, I think it makes sense to leave it the way it is.

Tests in `exhaustive_remote_user_type` mirror those from `exhaustive_user_type`. If an `exhaustive_user_type` test generates warnings, its counterpart in `exhaustive_remote_user_type` should generate the same or a similar warning. This wouldn't be easy to replicate if `exhaustive_remote_user_type` referred to tests in `user_types`, since the latter is a `should_pass` type test, whereas `exhaustive_user_type` is a `should_fail` type test. From yet another perspective, it makes sense to make `exhaustive_remote_user_type` and `exhaustive_user_type` work on the same types, but access them respectively as remote or local - it's just easier to reason about.

**Comment by erszcz** [comment]
I think this one is ready to merge, unless you have further comments.

Maybe the last commit introducing behaviour described by `test/known_problems/should_fail/exhaustive_map_variants.erl` is debatable, since AFAICT map typing is still not completely sound, and I'm not sure if you think map variants are a reasonable goal (I personally think they are). WDYT?

**Comment by zuiderkwast** [comment]
Yes, I will look through it again. I might have a few minor comment.

I think we should aim for handling map variants/unions.

File: Makefile
Comment (zuiderkwast) [inline Makefile:135]: The idea of not using the same `-I include -pa ebin` for the test programs as for the Gradualizer src itself was to make them independent. I think that was a good thing.

Is it the include or the ebin you need in the tests here?

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [inline src/gradualizer_lib.erl:198]: If there are multiple opaque types in a picked value, using the same variable name would imply all opaque values are equal, as in `{picked_tuple, _Opaque, _Opaque}`. That's why I suggested adding a unique integer to the name or use just a `_`. Not a big deal though since this only for an error message. There is a point in not flooding the atom table as well.

**Review by zuiderkwast — COMMENTED** [review]
I may accept it as is, if you reply to the comments. :-)

File: Makefile
Comment (erszcz) [reply Makefile:135]: Actually, the `debug_info`, as I think it's necessary to refer to remote types from `test_module` as in:

```
gradualizer -pa ./test_data -- test/should_fail/<test_module>.erl
```

File: test/known_problems/should_fail/exhaustive_map_variants.erl
Comment (erszcz) [reply test/known_problems/should_fail/exhaustive_map_variants.erl:16]: I haven't really looked for the reason of this not being detected - I'm aware that typing maps in general is not 100% correct yet, so it might be because of this PR or because of the existing code. I just wanted to add map variants for completeness sake, since ordinary tuples and records are already taken care of.

File: test/known_problems/should_fail/exhaustive_remote_map_variants.erl
Comment (erszcz) [reply test/known_problems/should_fail/exhaustive_remote_map_variants.erl:15]: I wanted to make it uniform with other examples from `exhaustive_remote_user_type` and `exhaustive_user_type` - i.e. that each local type tested for exhaustiveness is also accessed as a remote type. In the "worst" case (which is not that bad ;)) it'll just start passing as soon as the non-remote test is fixed.

Comment (zuiderkwast) [reply test/known_problems/should_fail/exhaustive_remote_map_variants.erl:15]: OK, fair enough.

File: Makefile
Comment (zuiderkwast) [reply Makefile:135]: Ah, so can we have a `TEST_ERLC_OPTS` instead or simply add `debug_info` on this command line?

File: src/gradualizer_lib.erl
Comment (erszcz) [reply src/gradualizer_lib.erl:198]: Indeed, this might be slightly misleading if more than one opaque type is used... I wanted to avoid passing extra mappings from types to names, but it shouldn't be hard to add - I'll do it in this PR.

File: Makefile
Comment (erszcz) [reply Makefile:135]: Yep. Ok, I'll take care of that 👍

File: src/gradualizer_lib.erl
Comment (zuiderkwast) [reply src/gradualizer_lib.erl:198]: Is there a need for passing around extra mappings? I was thinking just

```Erlang
list_to_atom("Opaque" ++ integer_to_list(erlang:unique_integer([positive])))
```

Numbers might look a bit random in this way but at least they would be unique.

Comment (zuiderkwast) [reply src/gradualizer_lib.erl:198]: > passing extra mappings from types to names

The same type used twice doesn't mean it's the same value, so I think it would actually be wrong to use the same variable twice in this case.

**Review by zuiderkwast — APPROVED** [review]
I'm happy now.

Do you wan to summarize a commit message for a squash merge?

**Comment by erszcz** [comment]
Maybe:

> Add non-trivial sum type exhaustiveness checking 
> 
> Enable exhaustiveness checking on sum types whose variants are tuples and records. Local and remote types, as well as (mutually) recursive types are supported. Map variant tests are added to known problems, but support is not present yet.

?

**Comment by zuiderkwast** [comment]
Awesome!

**Pull request merged**
